react-renderer/src/Evaluator.ts

```diff
@@ -208,23 +106,32 @@ const checkComp = (fn: string, args: ArgVal<number>[]) => {
  * TODO: update trans
  */
 export const evalShape = (
-  shapeExpr: IFGPI<number>,
-  trans: Translation,
-  varyingVars: VaryMap,
-  shapes: Shape[]
+    shapeExpr: IFGPI<Tensor>, // <number>?
+    trans: Translation,
+    varyingVars: VaryMap,
+    shapes: Shape[]
 ): [Shape[], Translation] => {
 
-  const [shapeType, propExprs] = shapeExpr.contents;
-  // Make sure all props are evaluated to values instead of shapes
-  const props = mapValues(propExprs, (prop: TagExpr<number>) =>
-    prop.tag === "OptEval"
-      ? (evalExpr(prop.contents, trans, varyingVars, false) as IVal<number>).contents
-      : prop.contents
-  );
+    const [shapeType, propExprs] = shapeExpr.contents;
+
+    // Make sure all props are evaluated to values instead of shapes
+    const props = mapValues(propExprs, (prop: TagExpr<Tensor>): Value<number> => {
+        if (prop.tag === "OptEval") {
+            // For display, evaluate expressions with autodiff types (incl. varying vars as AD types), then convert to numbers
+            // (The tradeoff for using autodiff types is that evaluating the display step will be a little slower, but then we won't have to write two versions of all computations)
```

Comment:
Will need to see profiling results to see the overhead of the `Tensor` => `number` step, otherwise okay with me.

Reply:
Here's a quick shot of three steps for the tree example (using tf.js on this branch). You can see that optimization takes most of the time and the conversion time seems negligible. Maybe this will change once we have compiled gradients. 

![image](https://user-images.githubusercontent.com/2762356/91885768-4fe56e00-ec56-11ea-86cc-a5d334331bb9.png)